pyrometheus/__init__.py

```diff
@@ -327,123 +326,232 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
+            volumetric) "bulk data", such as temperature, pressure, mass fractions,
+            etc. This parameter defaults to *actual numpy*, so it can be ignored
+            unless it is needed by the user (e.g. for GPU processing).
+
+        \"""
+
+        self.usr_np = usr_np
+        self.model_name = ${repr(sol.source)}
+        self.num_elements = ${sol.n_elements}
+        self.num_species = ${sol.n_species}
+        self.num_reactions = ${sol.n_reactions}
+        self.num_falloff = ${
+            sum(1 if isinstance(r, ct.FalloffReaction) else 0
+            for r in sol.reactions())}
 
-    def get_concentrations(self, rho, Y):
-        return self.iwts * rho * Y
+        self.one_atm = ${ct.one_atm}
+        self.gas_constant = ${ct.gas_constant}
+        self.big_number = 1.0e300
 
-    def get_mixture_specific_heat_cp_mass(self, T, Y):
-        return self.gas_constant * np.sum(
-            self.get_species_specific_heats_R(T)* Y * self.iwts)
+        self.species_names = ${sol.species_names}
+        self.species_indices = ${
+            dict([[sol.species_name(i), i]
+                for i in range(sol.n_species)])}
 
-    def get_mixture_specific_heat_cv_mass(self, T, Y):
-        cp0_R = self.get_species_specific_heats_R( T ) - 1.0
-        return self.gas_constant * np.sum(Y * cp0_R * self.iwts)
+        self.wts = ${str_np(sol.molecular_weights)}
+        self.iwts = 1/self.wts
 
-    def get_mixture_enthalpy_mass(self, T, Y):
-        h0_RT = self.get_species_enthalpies_RT( T )
-        return self.gas_constant * T * np.sum(Y * h0_RT * self.iwts)
+    def species_name(self, species_index):
+        return self.species_name[species_index]
 
-    def get_mixture_internal_energy_mass(self, T, Y):
-        e0_RT = self.get_species_enthalpies_RT( T ) - 1.0
-        return self.gas_constant * T * np.sum(Y * e0_RT * self.iwts)
+    def species_index(self, species_name):
+        return self.species_indices[species_name]
 
-    def get_species_specific_heats_R(self, T):
-        return np.array([
+    def get_specific_gas_constant(self, mass_fractions):
+        return self.gas_constant * np.dot(self.iwts, mass_fractions)
+
+    def get_density(self, p, temperature, mass_fractions):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        rt = self.gas_constant * temperature
+        return p * mmw / rt
+
+    def get_pressure(self, rho, temperature, mass_fractions):
+        mmw = self.get_mix_molecular_weight(mass_fractions)
+        rt = self.gas_constant * temperature
+        return rho * rt / mmw
+
+    def get_mix_molecular_weight(self, mass_fractions):
+        return 1/np.dot(self.iwts, mass_fractions)
+
+    def get_concentrations(self, rho, mass_fractions):
+        concs = self.iwts * rho * mass_fractions
+        zero = _pyro_zeros_like(concs[0])
+        for i, conc in enumerate(concs):
+            concs[i] = self.usr_np.where(concs[i] > 0, concs[i], zero)
```

Comment:
I found that it is messing up Jacobians computed using JAX. All Jacobian entries associated with species with zero concentration come back zero, even when they shouldn't be. (Note, even when a mass fraction is zero, its corresponding Jacobian entries can be non-zero). I got rid of it, and I was able to verify the Jacobian of get_net_production_rates against finite differences and analytical expressions.

I'd much prefer these sorts of squashing to be done outside of pyro, and this is one example of why it's not best practice. Hope this is clear. Let me know otherwise.

Reply:
So you allow negative concentrations in Pyro?

Reply:
Can we try this: (3fcdd54)?

Reply:
Agree with @ecisneros8, we should probably do this outside.